### PR TITLE
refactor(avoidance): use common safety checker

### PR DESCRIPTION
**Pull request by satoshi-ota**
### <samp>🤖 Generated by Copilot at 7d5d1f6</samp>

This pull request improves the safety check feature of the behavior path planner, which aims to avoid collisions with other objects by shifting lanes or decelerating. It updates the configuration file, the documentation, and the data structures for the new parameters and logic. It also refactors and simplifies the code of the avoidance module and the avoidance manager, and adds some utility functions for the safety check calculations.

Please approve :arrow_down: before this PR.
https://github.com/autowarefoundation/autoware_launch/pull/477

## Tests performed

- [x] [PASS TIER IV INTERNAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/020d66dd-eaae-5715-af8e-2d5731c6f0f5?project_id=prd_jt)

## Effects on system behavior

Nothing.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/utils/avoidance/utils.cpp [hunk 1]
```diff
@@ -1367,4 +1367,181 @@ AvoidLineArray combineRawShiftLinesWithUniqueCheck(
...
+      append_target_objects(check_lanes, data.other_objects);
+    }
+
+    if (p->check_unavoidable_object) {
+      append_target_objects(check_lanes, unavoidable_objects);
+    }
+  }
+
+  return target_objects;
+}
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/avoidance/utils.cpp:1546]: just my suggestion you can ignore but I thought `is_right` is better to be changed if my understand is correct .
```suggestion
std::vector<ExtendedPredictedObject> getSafetyCheckTargetObjects(
  const AvoidancePlanningData & data, const std::shared_ptr<const PlannerData> & planner_data,
  const std::shared_ptr<AvoidanceParameters> & parameters, const bool is_shifting_right)
{
  struct CheckCondition
  {
    bool should_check;
    enum CheckLane { RIGHT_LANE, LEFT_LANE, CURRENT_LANE } check_lane;
  };

  const auto & p = parameters;
  std::vector<ExtendedPredictedObject> target_objects;

  const auto unavoidable_objects = [&data]() {
    ObjectDataArray ret;
    std::for_each(data.target_objects.begin(), data.target_objects.end(), [&](const auto & object) {
      if (!object.is_avoidable) {
        ret.push_back(object);
      }
    });
    return ret;
  }();

  // clang-format off
  CheckCondition conditions[] = {
    {  is_shifting_right && p->check_shift_side_lane || !is_shifting_right && p->check_other_side_lane, CheckCondition::RIGHT_LANE },
    { !is_shifting_right && p->check_shift_side_lane ||  is_shifting_right && p->check_other_side_lane, CheckCondition::LEFT_LANE },
    { p->check_current_lane, CheckCondition::CURRENT_LANE },
  };
  // clang-format on

  for (const auto & condition : conditions) {
    if (!condition.should_check) {
      continue;
    }

    const auto check_lanes = condition.check_lane == CheckCondition::CURRENT_LANE
                               ? data.current_lanelets
                               : (condition.check_lane == CheckCondition::RIGHT_LANE
                                    ? getAdjacentLane(planner_data, p, true)
                                    : getAdjacentLane(planner_data, p, false));

    if (p->check_other_object) {
      for (const auto & object : data.other_objects) {
        if (isCentroidWithinLanelets(object.object, check_lanes)) {
          target_objects.push_back(utils::avoidance::transform(object.object, p));
        }
      }
    }

    if (p->check_unavoidable_object) {
      for (const auto & object : unavoidable_objects) {
        if (isCentroidWithinLanelets(object.object, check_lanes)) {
          target_objects.push_back(utils::avoidance::transform(object.object, p));
        }
      }
    }
  }

  return target_objects;
}
```

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM
Thank you so much for using common function. It helps when i make safety check class.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 2]
```diff
@@ -1832,371 +1833,55 @@ void AvoidanceModule::addReturnShiftLineFromEgo(AvoidLineArray & sl_candidates)
...
-    float travel_time = 0.0;
-    ret.emplace_back(path.points.front(), travel_time);
+  const auto ego_predicted_path =
+    utils::avoidance::convertToPredictedPath(shifted_path.path, planner_data_, parameters_);
 
-    for (size_t i = 1; i < path.points.size(); ++i) {
-      const auto & p1 = path.points.at(i - 1);
-      const auto & p2 = path.points.at(i);
+  const auto ego_idx = planner_data_->findEgoIndex(shifted_path.path.points);
+  const auto is_right_shift = [&]() -> std::optional<bool> {
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1848]: @satoshi-ota
Like always some small question happenst to me.
This `is_right_shift` is the flag to judge ego-vehicle is shifting and judge if safety check should be excuted based on the information i guess.
If so, only shifting to right direction is only left side running country like Japan?
So `is_shifting` or `is_avoiding` is more suitable I thought.

Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1848]: @kyoichi-sugahara Thanks for your question.

> This is_right_shift is the flag to judge ego-vehicle is shifting and judge if safety check should be excuted based on the information i guess.

Yes. If ego doesn't need to shift, the value will be `std::nullopt`.

> If so, only shifting to right direction is only left side running country like Japan?

No. I think there is a possiblity to meet the situation we have to shift right side on right-side traffic country. Additionaly, we sometimes met left shift situation in Odaiba.

Therefore, I think the variable `is_right_shift` needs to show following three case:

1. `is_right_shift: true` -> RIGHT side shift
2. `is_right_shift: false` -> LEFT side shift
3. `is_right_shift: std::nullopt`: NO shift lines exist

In conclusion, I suppose that it is better to use `is_right_shift`. What do you think?

File: planning/behavior_path_planner/src/utils/avoidance/utils.cpp (hunk 1 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/utils/avoidance/utils.cpp]: Fixed in https://github.com/autowarefoundation/autoware.universe/pull/4388/commits/e72a556715d9437fa6b0c47486470569a339f22d.

**Pull request merged**